sale_stock_picking_return_invoicing/models/stock_picking.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import api, fields, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    to_refund_lines = fields.Selection(selection=[
+        ('keep_line_value', 'Keep Line Value'),
+        ('to_refund_so', 'To Refund In Sales Order'),
+        ('no_refund_so', 'No Refund In Sales Order')],
+        string='Refund Options', default='keep_line_value',
+        help="This field allow modify 'to_refund_so' field value in all "
+             "stock moves from this picking after it has been confirmed."
+             "keep_line_value: keep the original value.\n"
+             "to_refund_so: set all stock moves to True value and recompute "
+             "delivered quantities in sale order.\n"
+             "no_refund_so: set all stock moves to False value and recompute "
+             "delivered quantities in sale order.",
+    )
+    location_usage = fields.Selection(
+        related='location_dest_id.usage',
+        readonly=True,
+    )
+
+    @api.multi
+    def _update_stock_moves(self):
+        for pick in self:
+            for move in pick.move_lines:
+                if pick.to_refund_lines == 'keep_line_value':
+                    move.to_refund_so = move.origin_to_refund_so
+                elif pick.to_refund_lines == 'to_refund_so':
+                    move.to_refund_so = True
+                elif pick.to_refund_lines == 'no_refund_so':
+                    move.to_refund_so = False
+
+    @api.multi
+    def set_delivered_qty(self):
+        so_lines = self.mapped(
+            'move_lines.procurement_id.sale_line_id').filtered(
+            lambda x: x.product_id.invoice_policy in ('order', 'delivery'))
+        for so_line in so_lines:
+            so_line.qty_delivered = so_line._get_delivered_qty()
```

Comment:
You should only update the lines that change its `to_refund_so` value.

Reply:
Is there a flag that can give me the state change at this point?